File: examples/experimental/scala-parallel-friend-recommendation/src/main/scala/Serving.scala
package org.apache.predictionio.examples.pfriendrecommendation

import org.apache.predictionio.controller.LServing
import org.apache.predictionio.controller.EmptyServingParams

class Serving
  extends LServing[Query, Double] {

  override
  def serve(query: Query, predictedResults: Seq[Double]): Double = {
    predictedResults.head
  }
}

File: examples/experimental/java-local-tutorial/src/main/java/recommendations/tutorial4/Serving.java
package org.apache.predictionio.examples.java.recommendations.tutorial4;

import org.apache.predictionio.controller.java.LJavaServing;
import org.apache.predictionio.controller.java.EmptyParams;
import java.lang.Iterable;

public class Serving extends LJavaServing<EmptyParams, Query, Float> {
  public Serving() {}

  public Float serve(Query query, Iterable<Float> predictions) {
    float sum = 0.0f;
    int count = 0;
    
    for (Float v: predictions) {
      if (!v.isNaN()) {
        sum += v;
        count += 1;
      }
    }
    return (count == 0) ? Float.NaN : sum / count;
  }
}

